manager/local.go

```diff
@@ -69,33 +73,90 @@ func (m *Manager) prepareGitRepoPath(catalog model.Catalog, update bool, catalog
 	repoPath := path.Join(m.cacheRoot, catalog.EnvironmentId, repoBranchHash)
 
 	if err := os.MkdirAll(repoPath, 0755); err != nil {
-		return "", "", catalogType, err
+		return "", "", catalogType, errors.Wrap(err, "mkdir failed")
 	}
 
 	empty, err := dirEmpty(repoPath)
 	if err != nil {
-		return "", "", catalogType, err
+		return "", "", catalogType, errors.Wrap(err, "Empty directory check failed")
 	}
 
 	if empty {
 		if err = git.Clone(repoPath, catalog.URL, branch); err != nil {
-			return "", "", catalogType, err
+			return "", "", catalogType, errors.Wrap(err, "Clone failed")
 		}
 	} else {
 		if update {
-			if git.RemoteShaChanged(catalog.URL, catalog.Branch, catalog.Commit, m.uuid) {
+			changed, err := m.remoteShaChanged(catalog.URL, catalog.Branch, catalog.Commit, m.uuid)
+			if err != nil {
+				return "", "", catalogType, errors.Wrap(err, "Remote commit check failed")
+			}
+			if changed {
 				if err = git.Update(repoPath, branch); err != nil {
-					// Ignore error unless running in strict mode
-					if m.strict {
-						return "", "", catalogType, err
-					}
-					log.Errorf("Failed to update existing repo cache: %v", err)
+					return "", "", catalogType, errors.Wrap(err, "Update failed")
 				}
 				log.Debugf("catalog-service: updated catalog '%v'", catalog.Name)
 			}
 		}
 	}
 
 	commit, err := git.HeadCommit(repoPath)
+	if err != nil {
+		err = errors.Wrap(err, "Retrieving head commit failed")
+	}
 	return repoPath, commit, catalogType, err
 }
+
+func formatGitURL(endpoint, branch string) string {
+	formattedURL := ""
+	if u, err := url.Parse(endpoint); err == nil {
+		pathParts := strings.Split(u.Path, "/")
+		switch strings.Split(u.Host, ":")[0] {
+		case "github.com":
+			if len(pathParts) >= 3 {
+				org := pathParts[1]
+				repo := strings.TrimSuffix(pathParts[2], ".git")
+				formattedURL = fmt.Sprintf("https://api.github.com/repos/%s/%s/commits/%s", org, repo, branch)
+			}
+		case "git.rancher.io":
+			repo := strings.TrimSuffix(pathParts[1], ".git")
+			u.Path = fmt.Sprintf("/repos/%s/commits/%s", repo, branch)
+			formattedURL = u.String()
+		}
+	}
+	return formattedURL
+}
+
+func (m *Manager) remoteShaChanged(repoURL, branch, sha, uuid string) (bool, error) {
+	formattedURL := formatGitURL(repoURL, branch)
+
+	if formattedURL == "" {
+		return true, nil
+	}
+
+	req, err := http.NewRequest("GET", formattedURL, nil)
+	if err != nil {
+		log.Warnf("Problem creating request to check git remote sha of repo [%v]: %v", repoURL, err)
+		return true, nil
+	}
+	req.Header.Set("Accept", "application/vnd.github.chitauri-preview+sha")
+	req.Header.Set("If-None-Match", fmt.Sprintf("\"%s\"", sha))
+	if uuid != "" {
+		req.Header.Set("X-Install-Uuid", uuid)
+	}
+	res, err := m.httpClient.Do(req)
+	if err != nil {
+		// Return timeout errors so caller can decide whether or not to proceed with updating the repo
+		if uErr, ok := err.(*url.Error); ok && uErr.Timeout() {
```

Comment:
I chose to only bubble up the error if it was explicitly a timeout error. The goal is to prevent the next step of doing a git fetch or clone from also timing out (which will take much logner). For non-timeout errors, the thinking is that the clone/fetch will either fail the same way or it wont.